pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -155,21 +160,32 @@ public PinotHelixResourceManager(@Nonnull String zkURL, @Nonnull String helixClu
     _isSingleTenantCluster = isSingleTenantCluster;
     _enableBatchMessageMode = enableBatchMessageMode;
     _allowHLCTables = allowHLCTables;
+    _helixInstanceType = helixInstanceType;
+
+    _helixZkManager =
+        HelixManagerFactory.getZKHelixManager(_helixClusterName, _instanceId, _helixInstanceType, _helixZkURL);
+    // LeadControllerManager needs to be initialized before registering as Helix participant.
+    _leadControllerManager = new LeadControllerManager(_instanceId, _helixZkManager);
   }
 
-  public PinotHelixResourceManager(@Nonnull ControllerConf controllerConf) {
-    this(controllerConf.getZkStr(), controllerConf.getHelixClusterName(),
-        CommonConstants.Helix.PREFIX_OF_CONTROLLER_INSTANCE + controllerConf.getControllerHost() + "_" + controllerConf
-            .getControllerPort(), controllerConf.getDataDir(), controllerConf.getExternalViewOnlineToOfflineTimeout(),
+  // This constructor should only be explicitly called by {@link ControllerStarter}.
+  public PinotHelixResourceManager(@Nonnull ControllerConf controllerConf, InstanceType instanceType) {
+    this(controllerConf.getZkStr(), controllerConf.getHelixClusterName(), LeadControllerUtils
+            .generateControllerParticipantId(controllerConf.getControllerHost(), controllerConf.getControllerPort()),
+        controllerConf.getDataDir(), controllerConf.getExternalViewOnlineToOfflineTimeout(),
         controllerConf.tenantIsolationEnabled(), controllerConf.getEnableBatchMessageMode(),
-        controllerConf.getHLCTablesAllowed());
+        controllerConf.getHLCTablesAllowed(), instanceType);
+  }
+
+  public PinotHelixResourceManager(@Nonnull ControllerConf controllerConf) {
+    this(controllerConf, InstanceType.ADMINISTRATOR);
```

Comment:
Don't change the default behavior, join as PARTICIPANT by default

Reply:
Removed the instance type from resource manager and put the logic to the caller instead.